crates/core/src/replication/connection.rs

```diff
@@ -92,16 +148,50 @@ impl RemoteConnection {
     }
 
     pub(self) fn should_execute_local(&self, stmts: &[parser::Statement]) -> bool {
-        let is_read_only = stmts.iter().all(|s| s.is_read_only());
+        let mut inner = self.inner.lock();
+
+        let predicted_end_state = predict_final_state(inner.state, stmts.iter());
+
+        match (inner.state, predicted_end_state) {
```

Comment:
This PR is in draft because I think we really should add tests for this state machine, as there are likely some bugs.